vavr/src/main/java/io/vavr/collection/Array.java

```diff
@@ -984,21 +984,63 @@ public int length() {
 
     @Override
     public Array<Array<T>> permutations() {
-        if (isEmpty()) {
-            return empty();
-        } else if (delegate.length == 1) {
-            return of(this);
+        final Array<T> src = distinct();
+        if (src.isEmpty()) {
+            return Array.empty();
+        } else if (src.size() == 1) {
+            return Array.of(src);
         } else {
-            Array<Array<T>> results = empty();
-            for (T t : distinct()) {
-                for (Array<T> ts : remove(t).permutations()) {
-                    results = results.append(of(t).appendAll(ts));
+            java.util.List<Array<T>> result = new java.util.ArrayList<>();
+            int[] pi = new int[src.size()];
+
+            // see Narayana algo
+            // https://en.wikipedia.org/wiki/Permutation#Generation_in_lexicographic_order
+
+            for (int i = 0; i < pi.length; i++) {
+                pi[i] = i;
+            }
+
+            int i, j;
+            while (true) {
+                Object[] next = new Object[src.size()];
+                for (int k = 0; k < next.length; k++) {
+                    next[k] = src.delegate[pi[k]];
+                }
+                result.add(wrap(next));
+
+                i = pi.length - 2;
+                while (i >= 0 && pi[i] > pi[i + 1]) {
+                    i--;
+                }
+                if (i < 0) {
+                    break;
+                }
+
+                j = pi.length - 1;
+                while (pi[i] > pi[j]) {
+                    j--;
+                }
+                swapIndices(pi, i, j);
+
+                i++;
+                j = pi.length - 1;
+                while (i < j) {
+                    swapIndices(pi, i, j);
+                    i++;
+                    j--;
                 }
             }
```

Comment:
**Just a thought, no requirement to change it:** I think the algorithm could be useful for all IndexSeq impls. I think we only need to make the factory methods `wrap` and `ofAll` generic. But I see that the other IndexedSeq impls do not operate on Object[], so maybe it is not possible to make it generic for impls though.